stackslib/src/chainstate/burn/atc.rs

```diff
@@ -1401,3 +1158,314 @@ pub(crate) const ATC_LOOKUP: [AtcRational; 1024] = [
     AtcRational(Uint256([18041977930440052, 0, 0, 0])),
     AtcRational(Uint256([17835588001385282, 0, 0, 0])),
 ];
+
+#[cfg(test)]
+mod test {
+    use stacks_common::util::hash::to_hex;
+    use stacks_common::util::uint::Uint256;
+
+    use crate::chainstate::burn::atc::AtcRational;
+    use crate::chainstate::burn::BlockSnapshot;
+    use crate::stacks_common::util::uint::BitArray;
+
+    impl AtcRational {
+        /// Convert to f64, and panic on conversion failure
+        pub fn to_f64(&self) -> f64 {
+            let ipart = self.ipart() as f64;
+            let fpart = self.0.low_u64() as f64;
+            ipart + (fpart / (u64::MAX as f64))
+        }
+
+        /// Convert from f64 between 0 and 1, panicking on conversion failure.  Scales up the f64 so that its
+        /// fractional parts reside in the lower 64 bits of the AtcRational.
+        pub fn from_f64_unit(value: f64) -> Self {
+            if value < 0.0 || value >= 1.0 {
+                panic!("only usable for values in [0.0, 1.0) range");
+            }
+
+            // NOTE: this only changes the exponent, not the mantissa.
+            // Moreover, u128::from(u64::MAX) + 1 has f64 representation 0x43f0000000000000, so these conversions are safe.
+            let scaled_value = value * ((u128::from(u64::MAX) + 1) as f64);
+
+            // this is safe, because 0.0 <= value < 1.0, so scaled_value <= u64::MAX
+            let value_u64 = scaled_value as u64;
+            Self(Uint256::from_u64(value_u64))
+        }
+    }
+
+    #[test]
+    fn test_atc_rational() {
+        // zero
+        assert_eq!(AtcRational::zero().into_inner(), Uint256::from_u64(0));
+
+        // one
+        assert_eq!(AtcRational::one().into_inner(), Uint256::one() << 64);
+
+        // one_sup
+        assert_eq!(
+            AtcRational::one_sup().into_inner(),
+            (Uint256::one() << 64) - Uint256::from_u64(1)
+        );
+
+        // max
+        assert_eq!(
+            AtcRational::max().into_inner(),
+            (Uint256::from_u64(u64::MAX) << 64) | Uint256::from_u64(u64::MAX)
+        );
+
+        // ipart
+        assert_eq!(AtcRational::one().ipart(), 1);
+        assert_eq!(AtcRational::frac(1, 2).ipart(), 0);
+        assert_eq!(AtcRational::frac(3, 2).ipart(), 1);
+        assert_eq!(AtcRational::frac(4, 2).ipart(), 2);
+        assert_eq!(AtcRational::frac(9999, 10000).ipart(), 0);
+
+        // to_f64
+        assert_eq!(AtcRational::one().to_f64(), 1.0);
+        assert_eq!(AtcRational::zero().to_f64(), 0.0);
+        assert_eq!(AtcRational::frac(1, 2).to_f64(), 0.5);
+        assert_eq!(AtcRational::frac(1, 32).to_f64(), 0.03125);
+
+        // from_f64_unit
+        assert_eq!(AtcRational::from_f64_unit(0.0), AtcRational::zero());
+        assert_eq!(AtcRational::from_f64_unit(0.5), AtcRational::frac(1, 2));
+        assert_eq!(
+            AtcRational::from_f64_unit(0.03125),
+            AtcRational::frac(1, 32)
+        );
+
```

Comment:
I think the important properties to test here are something like:

```
fn check_add(num_1: u64, den_1: u64, num_2: u64, den_2: u64) {
    assert_eq!(
        (AtcRational::frac(num_1, den_1) + AtcRational::frac(num_2, den_2)).to_f64(),
        num_1 as f64 / den_1 as f64 + num_2 as f64 / den_2 as f64
    );
    assert_eq!(
        (AtcRational::frac(num_1, den_1) + AtcRational::frac(num_2, den_2)),
        AtcRational::from_f64(num_1 as f64 / den_1 as f64 + num_2 as f64 / den_2 as f64)
    );
}
```

And a similar property for multiplication. And then that function needs to be invoked on a sufficient range of the space to be convincing that the math is correct. Right now, its just not obvious that AtcRational leads to sane mathematics. Like, it looks correct to me, but this kind of math library can have some nasty corner cases.

Reply:
I don't think that second assert is necessary, because we never construct `AtcRational` from `f64` outside of tests (and even then, we only do so for the domain of `[0.0, 1.0)` via `from_f64_unit()`).

I'll add the first assert though, but at a granularity that makes sense.  The trouble with comparing derived `f64`s is that they're lossy.

Reply:
Added test coverage for representing the range `[0.0, 2.0)` because while we never construct an `AtcRational` outside the range `[0.0, 1.0)`, we will take such an `AtcRational` and scale it up to index into the lookup table (i.e. multiply by 1024).